cpmpy/expressions/globalconstraints.py

```diff
@@ -312,6 +312,146 @@ def value(self):
+        for s, v, e in transitions:
+            self.mapping[(s, v)] = e
+
+    def transition_to_layer_representation(self):
+        arr, transitions = self.args
+        nodes_by_level = [[self.root_node]]
+        transitions_by_level = []
+        tran = transitions
+        for i in range(len(arr)):
+            nodes_by_level.append([])
+            transitions_by_level.append([])
+            remaining_tran = []
+            for ns, v, ne in tran:
+                if ns in nodes_by_level[i]:
+                    if ne not in nodes_by_level[i + 1]:
+                        nodes_by_level[i + 1].append(ne)
+                    transitions_by_level[i].append((ns, v, ne))
+                else:
+                    remaining_tran.append((ns, v, ne))
+            tran = remaining_tran
+        return nodes_by_level, transitions_by_level
+
+    # auxillary method to transform into layered representation (gather all the node by node-layers)
+    def normalize_layer_representation(self, nodes_by_level, transitions_by_level):
+        nb_nodes_by_level = [len(x) for x in nodes_by_level]
+        num_mapping = {}
+        for lvl in nodes_by_level:
+            for i in range(len(lvl)):
+                num_mapping[lvl[i]] = i
+        transitions_by_level_normalized = [[[num_mapping[n_in], v, num_mapping[n_out]] for n_in, v, n_out in lvl] for
+                                           lvl in
+                                           transitions_by_level]
+        return nb_nodes_by_level, num_mapping, transitions_by_level_normalized
+
+
+    def decompose(self):
+        # Table decomposition (not by decomposition of the mdd into one big table, but by having transition table for
+        # each state and auxiliary variables for the nodes. Similar to decomposition of regular into table,
+        # but with one table for each layer
+        arr, _ = self.args
+        lb = [x.lb for x in arr]
+        ub = [x.ub for x in arr]
+        nbl, tbl = self.transition_to_layer_representation()
+        nb_nodes_by_level, num_mapping, transitions_by_level_normalized = self.normalize_layer_representation(nbl, tbl)
+        if len(transitions_by_level_normalized) > 2:
+            # decomposition with multiple transitions table and aux variables for the nodes
+            aux = [intvar(0, nb_nodes) for nb_nodes in nb_nodes_by_level[1:]]
+            for i in range(len(arr)):
+                transition_dummy = [[num_mapping[n], v, nb_nodes_by_level[i+1]] for n in nbl[i] for v in range(lb[i], ub[i] + 1) if
+                            (n, v) not in self.mapping]
+                if i != 0:
+                    transition_dummy += [[nb_nodes_by_level[i], v, nb_nodes_by_level[i+1]] for v in range(lb[i], ub[i] + 1)]
+                transitions_by_level_normalized[i] = transitions_by_level_normalized[i] + transition_dummy
+            tab_first = [x[1:] for x in transitions_by_level_normalized[0]]  # optimization for first level (one node)
```

Comment:
explain that we don't need to know the "start state" for the first transition, as it is deterministic.